### DISCUSSION TITLE
Restart a closed application after the installation completed succesfully

### DISCUSSION THREAD

**Discussion by hotzenwalder** [PSAppDeployToolkit/PSAppDeployToolkit#848] [Ideas]
For some installations we need to close running apps like Teams or Outlook. The closed apps are not restarted after a succesful installation. It would be nice if we had an option to specify which processes or apps we'd like to restart after a succesful installation if they are not running. 

Some apps have to be started in the user context, so it must also be possible to specify if the process or application must be executed as a user or as system and if any parameters are needed to restart the application.

We now use this code

>If($StartApp -eq "true" -and $AppsClosed -eq "true" -and $StartCMD -ne $null -and $StartAs -ne $null){
    If($StartAs -eq "user"){
        If ($StartParameters){
                    Execute-ProcessAsUser -Path "$StartCMD" -Parameters "$StartParameters"
                }
                else{
                    Execute-ProcessAsUser -Path "$StartCMD"
                }
            }
            else{
                If ($StartParameters){
                    Execute-Process -Path "$StartCMD" -Parameters "$StartParameters"
                }        
                else{
                    Execute-Process -Path "$StartCMD"
                }  
            }
        }

For the parameters we use these variables
> $StartApp = "true" ## true / false
$StartAs = "system"  ## system / user
$StartCMD = "C:\Program Files\Google\Chrome\Application\chrome.exe" ## path to the application exe
$StartParameters = "" ## Extra parameters needed to start

To determine if any processes were closed and need to be restarted we added this code

>$AppsClosed = $false    
                $CheckProcess =$CloseAppProces.Split(",")   
                $Array = [System.Collections.ArrayList]@()
                Foreach ($item in $CheckProcess) {
                    $targetprocesses = @(Get-WmiObject -Query "Select * FROM Win32_Process WHERE Name='$item.exe'" -ErrorAction SilentlyContinue)
                    If($targetprocesses){
                        $Data = New-Object PSObject
                        $Data | Add-Member -MemberType NoteProperty -Name "Process" -Value $targetprocesses.name
                        $Data | Add-Member -MemberType NoteProperty -Name "User" -Value $targetprocesses.GetOwner().User     
                        $Array.Add($Data) | Out-Null
                    }
                }
                if ($array.Count -ge 1) {
                    $AppsClosed = $true
                } else {
                   $AppsClosed = $false
                }

$Checkprocess uses $CloseAppProces which looks like this

>$CloseAppProces = "googleupdate,chrome"

Needs some work and some code is probably not optimal, but I hope you get the idea. We added this code because there doesn't seem to be a way to detect if any process was closed during the installation or are we missing something?

**Comment by mjr4077au** [comment]
The problem with trying to automate this is that the apps won't ever return to the exact state that they were in. I think this is better left manual for the user to do.

**Comment by DanGough** [comment]
This is actually a common request and it's not unsolvable. Many apps will start back up how you left them (e.g. Chrome, Notepad++, VSCode), and even if they don't, at least they get fired back up again.

There are a few things to take note of however:

- Packagers should be able to specify a separate list of apps to restart, as this will likely be a shorter list than the ones that need to be closed.
- Starting arguments can be captured from the running processes before  they are terminated.
- Because multiple users can be logged on, apps should be relaunched in the correct user context.
- To avoid killing services and restarting them as regular processes, the function could have separate support for stopping/starting services

**Reply by mjr4077au** [reply]
I've created https://github.com/PSAppDeployToolkit/PSAppDeployToolkit/issues/1072 so we can track the potential implementation of this.

**Reply by Shaun-Mullins** [reply]
I agree, if you are asking to close applications that are running then you should be able to specify if you would like to start a process if it was originally started. But I would only want to start a process that was already running, I would not want to start a process that was not originally started.  As the user might have the software installed as a requirement, but they have disabled the application to auto start.

I would like to specify what apps to open which might be different from what was closed.  So I would not like to see this Option a True / False one.

So where we have 

AppProcessesToClose ('Greenshot','App2','App3')

We could have 

AppProcessesToOpenIfClosed ('Greenshot')

Asking a user via a dialog box to open something, is not good customer service.  If the process was already running it should automatically be started after the installation.  If the user misses the dialog box or forgets, in today's world where they never restart their computer, they just do sleep or hibernate.  It would be better to put the computer back to the same state it was before the installation.
